File: prototype/php/src/app/Models/Payout.php
<?php

declare(strict_types=1);

namespace App\Models;

use App\Domain\Money\Money;
use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Override;

/**
 * @property-read Seller $seller
 */
#[Fillable(['seller_id', 'period_start', 'period_end', 'amount_cents', 'paid_at'])]
class Payout extends Model
{
    /**
     * @return array<string, string>
     */
    #[Override]
    protected function casts(): array
    {
        return [
            'period_start' => 'date',
            'period_end' => 'date',
            'amount_cents' => 'integer',
            'paid_at' => 'datetime',
        ];
    }

    /** @return BelongsTo<Seller, $this> */
    public function seller(): BelongsTo
    {
        return $this->belongsTo(Seller::class);
    }

    /** @return HasMany<LedgerEntry, $this> */
    public function ledgerEntries(): HasMany
    {
        return $this->hasMany(LedgerEntry::class);
    }

    public function amount(): Money
    {
        return Money::fromCents($this->amount_cents);
    }
}

File: prototype/rails/src/app/models/payout.rb
class Payout < ApplicationRecord
  prefixed_id :pyt

  belongs_to :seller
  has_many :ledger_entries, dependent: :nullify

  scope :for_seller, ->(seller_id) { where(seller_id: seller_id) if seller_id.present? }

  # The weekly settlement: one row per seller holding money released and not
  # yet sent, for the Monday-to-Sunday week that ended before as_of.
  def self.run_weekly(as_of: Time.current)
    period = PayoutPeriod.ending_before(as_of)

    Story.tell("payout.run", "settling the week that ended #{period.last_day}",
      period_start: period.first_day.to_s, period_end: period.last_day.to_s) do |story|
      payouts = transaction do
        LedgerEntry.occurred_by(period.ends_at).balances_by_seller
                   .select { |_seller_id, balance| balance.payable? }
                   .map { |seller_id, balance| settle(seller_id, balance.available, period, as_of) }
      end

      story.did("settled the week that ended #{period.last_day}",
        period_start: period.first_day.to_s, period_end: period.last_day.to_s,
        payout_count: payouts.size, total_cents: payouts.sum(&:amount_cents))

      payouts
    end
  end

  private_class_method def self.settle(seller_id, available, period, as_of)
    Story.tell("payout.pay", "paying a seller what the week released",
      seller_id: seller_id, amount_cents: available.cents) do |story|
      payout = pay(seller_id, available, period, as_of)

      story.did("paid the seller", seller_id: seller_id, payout_id: payout.id, amount_cents: payout.amount_cents)

      payout
    end
  end

  private_class_method def self.pay(seller_id, available, period, as_of)
    payout = create!(
      seller_id: seller_id,
      period_start: period.first_day, period_end: period.last_day,
      amount_cents: available.cents, paid_at: as_of
    )

    # Dated at the close of the period it settles, so a second run of the same
    # period already counts the money as sent and pays nothing.
    LedgerEntry.pay_out(payout, at: period.ends_at)

    payout
  end

  def amount
    Money.from_cents(amount_cents)
  end
end
